In [19]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats

### 2. Given the following confusion matrix, evaluate (by hand) the model's performance.

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |


- In the context of this problem, what is a false positive?
- In the context of this problem, what is a false negative?
- How would you describe this model?

##### - Positive case: Dogs
##### - Negative case: Cats

##### - False positive: predicting dogs and actuals are cats (13)
##### - False negative: predicting cats and actuals are dogs (7)

In [27]:
# - Description of the model: this model is a DOG classifier 
# because DOGS are the positive prediction

tp = true_positive = 46
tn = true_negative = 34
fp = false_positive = 13
fn = false_negative = 7

accuracy = (tp + tn) / (tp + tn + fp + fn)
recall = tp / (tp + fn)
precision = tp / (tp + fp)

print(f'Accuracy: {accuracy}\n')
print(f'Recall: {round(recall, 2)}\n')
print(f'Precision: {round(precision, 2)}')

Accuracy: 0.8

Recall: 0.87

Precision: 0.78


### 3. You are working as a datascientist for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

##      a.

In [12]:
# Getting to know the content of the DataFrame

c3 = pd.read_csv('c3.csv')
c3

actual     model1     model2     model3
0    No Defect  No Defect     Defect  No Defect
1    No Defect  No Defect     Defect     Defect
2    No Defect  No Defect     Defect  No Defect
3    No Defect     Defect     Defect     Defect
4    No Defect  No Defect     Defect  No Defect
..         ...        ...        ...        ...
195  No Defect  No Defect     Defect     Defect
196     Defect     Defect  No Defect  No Defect
197  No Defect  No Defect  No Defect  No Defect
198  No Defect  No Defect     Defect     Defect
199  No Defect  No Defect  No Defect     Defect

[200 rows x 4 columns]

In [31]:
# Looking for null values and their count

c3.isna().value_counts()

actual  model1  model2  model3
False   False   False   False     200
dtype: int64

In [28]:
# Discovering the count of each variable in actual data

c3.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

**Summary for myself:**
- Mode or baseline = 184 (no defects, the number that is repeated the most)
- type of model = a defect classifier (we are looking for defects)
- metric = Recall (recall measure the misses and I don't want to miss the defects)

##### - Since the goal is to identify as many defects as possible, defect is the classifier 
##### and recall the metric because it checks for how many defects are actually flagged

In [41]:
pd.crosstab(c3.actual, c3.model1)

model1     Defect  No Defect
actual                      
Defect          8          8
No Defect       2        182

In [42]:
pd.crosstab(c3.actual, c3.model2)

model2     Defect  No Defect
actual                      
Defect          9          7
No Defect      81        103

In [43]:
pd.crosstab(c3.actual, c3.model3)

model3     Defect  No Defect
actual                      
Defect         13          3
No Defect      86         98

In [48]:
recall_1 = 8/(8+7)

recall_2 = 9/(9+3)

recall_3 = 13/(13+3)


print(f'Recall_1: {round(recall_1, 2)}\n')
print(f'Recall_1: {recall_2}\n')
print(f'Recall_1: {round(recall_3, 2)}\n')

Recall_1: 0.53

Recall_1: 0.75

Recall_1: 0.81



##### The recall rate is higher for model3. I suggest the internal team choose it

### b.

### 4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

#### a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

#### b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

#### c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

### 5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.